### ISSUE TITLE
insoluble nann problem in training (loss = nan)

### ISSUE THREAD

**Issue by yusufislam98** [amdegroot/ssd.pytorch#373]
I've read all the other nann isues and apply all solution methods that mainly;
1)decreasing learning rate as 1e-4, 1e-5, 1e-6, 1e-7 also 0
2)changing momentum value as 0.8, 0.99
3)using coco dataset and my own dataset
4)1,2,4,6,16,32 batch sizes
5)training on GPU,CPU and Colab
6) rebuilt the SSD algorithm
But, there was no change when all of combinations of this solutions apllied and algorithm gives nan value after about 50,60,70 iterations...

How can I overcome to this issue....

**Comment by MIXALER** [comment]
the same question!!!

**Comment by indrajitsg** [comment]
Check your localizer loss, I think it is touching infinity for you. Check this [issue](https://github.com/amdegroot/ssd.pytorch/issues/162).

**Comment by isthatyoung** [comment]
I met this problem these day but solved, one hint, you could first look at you dataset, double check whether it has 'dirty' data likes no annotations, xmin >= xmax or ymin >= ymax. My dataset has these kind of 'dirty' data which cause the loss became nan or inf. Then you could use smaller learning rate to train you SSD.

**Comment by zhaohao0404** [comment]
> I've read all the other nann isues and apply all solution methods that mainly;
> 1)decreasing learning rate as 1e-4, 1e-5, 1e-6, 1e-7 also 0
> 2)changing momentum value as 0.8, 0.99
> 3)using coco dataset and my own dataset
> 4)1,2,4,6,16,32 batch sizes
> 5)training on GPU,CPU and Colab
> 6) rebuilt the SSD algorithm
> But, there was no change when all of combinations of this solutions apllied and algorithm gives nan value after about 50,60,70 iterations...
> 
> How can I overcome to this issue....

sometimes the key part may locate at initialize function. you can have a try with different ones.

**Comment by chirico85** [comment]
Hi, 
I had the same problem right from training iteration < 100 on cpu/gpu. 
My steps to find a reason for loss = nan were:
- run training on cpu only
- check for 'dirty' data
- reduce number of own dataset and keep only images of equal sizes  --> resulted in loss = nan after training iteration > 300
- reduce lr from 0.08 to 0.05 --> resulted in loss != nan 
- increased number of dataset --> resulted in loss != nan 

I am going to check if that works on gpu as well and report.

**Comment by chirico85** [comment]
> Hi, I had the same problem right from training iteration < 100 on cpu/gpu. My steps to find a reason for loss = nan were:
> 
> * run training on cpu only
> * check for 'dirty' data
> * reduce number of own dataset and keep only images of equal sizes  --> resulted in loss = nan after training iteration > 300
> * reduce lr from 0.08 to 0.05 --> resulted in loss != nan
> * increased number of dataset --> resulted in loss != nan
> 
> I am going to check if that works on gpu as well and report.

Training on GPU with same conditions provides loss=nan.
Can someone help?

**Comment by zuliani99** [comment]
@isthatyoung so swapping the coordinates of the boxes in the following cases should solve the problem?

I mean if I manage the list of coordinates as follow `['xmin', 'ymin', 'xmax', 'ymax']`, I should swap the coordinates that satisfy these two cases:

```python
if bndbox[0] >= bndbox[2]:
    temp = bndbox[0]
    bndbox[0] = bndbox[2]
    bndbox[2] = temp
                
if  bndbox[1] >= bndbox[3]:
    temp = bndbox[1]
    bndbox[1] = bndbox[3]
    bndbox[3] = temp
```
